packages/template/src/lib/stack-app/apps/implementations/client-app-impl.ts

```diff
@@ -437,17 +523,137 @@ export class _StackClientAppImplIncomplete<HasTokenStore extends boolean, Projec
+    runAsynchronously(async () => {
+      const domain = await this._fetchTrustedParentDomain();
+      const previousDomain = this._lastCustomRefreshCookieDomain;
+      const setOrDeleteCookieWrapper = context === "browser"
+        ? async (...args: Parameters<typeof setOrDeleteCookieClient>) => setOrDeleteCookieClient(...args)
+        : setOrDeleteCookie;
+
+      const deleteCookie = async (targetDomain: string) => {
+        const name = this._getCustomRefreshCookieName(targetDomain);
+        await setOrDeleteCookieWrapper(name, null, { domain: targetDomain, noOpIfServerComponent: true });
+      };
+      const setCookie = async (targetDomain: string, value: string) => {
+        const name = this._getCustomRefreshCookieName(targetDomain);
+        await setOrDeleteCookieWrapper(name, value, { maxAge: 60 * 60 * 24 * 365, domain: targetDomain, noOpIfServerComponent: true });
+      };
+
+      if (!domain) {
+        if (previousDomain) {
+          await deleteCookie(previousDomain);
+          this._lastCustomRefreshCookieDomain = null;
+        }
+        return;
+      }
+
+      if (previousDomain && previousDomain !== domain) {
+        await deleteCookie(previousDomain);
+      }
+
+      if (refreshToken && updatedAt !== null) {
+        if (this._lastCustomRefreshCookieUpdatedAt !== null && updatedAt < this._lastCustomRefreshCookieUpdatedAt) {
+          // Ignore stale writes
+          return;
+        }
+        const value = this._formatRefreshCookieValue(refreshToken, updatedAt);
+        await setCookie(domain, value);
+        this._lastCustomRefreshCookieDomain = domain;
+        this._lastCustomRefreshCookieUpdatedAt = updatedAt;
+      } else {
+        await deleteCookie(domain);
+        this._lastCustomRefreshCookieDomain = null;
+        this._lastCustomRefreshCookieUpdatedAt = null;
+      }
+    });
+  }
+  private async _fetchTrustedParentDomain(): Promise<string | null> {
+    const project = Result.orThrow(await this._interface.getClientProject());
+    const domains = project.config.domains;
+    const direct = new Set<string>();
+    const wildcard = new Set<string>();
+    for (const domain of domains) {
+      const withoutProtocol = domain.domain.trim().replace(/^https?:\/\//, "");
+      const host = withoutProtocol.split("/")[0]?.toLowerCase();
+      if (host.startsWith("*.")) {
+        const candidate = host.slice(2);
+        if (candidate && !candidate.includes("*")) {
+          wildcard.add(candidate);
+        }
+      } else if (!host.includes("*")) {
+        direct.add(host);
+      }
+    }
+    const candidates = Array.from(direct).filter((domain) => wildcard.has(domain));
+    candidates.sort((a, b) => {
+      const aParts = a.split(".").length;
+      const bParts = b.split(".").length;
+      if (aParts !== bParts) {
+        return aParts - bParts;
+      }
+      return stringCompare(a, b);
+    });
+    const selected = candidates[0] ?? null;
+    return selected;
+
+  }
```

Comment:
i think a simpler implementation of this function takes the current domain as an argument, and then iterates by going up one level in the domain each; for each parent domain, it checks whether both the parent domain itself *and* `**.parent-domain` are trusted domains




```suggestion
  private async _fetchTrustedParentDomain(currentDomain?: string): Promise<string | null> {
    const project = Result.orThrow(await this._interface.getClientProject());
    const domains = project.config.domains.map(d => d.domain.trim().replace(/^https?:\/\//, "").split("/")[0]?.toLowerCase();

    const parts = currentDomain.split('.');
    for (let i = part.length - 2; i >= 0; i--) {
      const parentDomain = parts.slice(i).join('.');
      if (domains.includes(parentDomain) && trustedWildcards.includes("**." + parentDomain)) {
        return parentDomain;
      }
    }
    
    return null;
  }
```